### ISSUE TITLE
Overridable parent element for search results

### ISSUE THREAD

**Issue by m-gagne** [EvanBoyle/AzSearch.js#33]
Currently in search results, each item is wrapped in a div. This makes it impossible to use a table as the parent result element and row/columns for the results. Would require optionally passing in the element type to use in `Results.tsx: return()` (see: https://github.com/EvanBoyle/AzSearch.js/blob/master/src/components/Results.tsx#L26)

**Comment by EvanBoyle** [comment]
Do you have a proposal for what the API would look like? Do you think this would be a common ask?

I'm not sure that this is a scenario that would be easily accommodated with a single api parameter. If you're going to use a table, you probably want to insert a table header is well. At that point we're talking about enabling a greater level of customization with rendering. 

I had some thoughts about ways to allow custom rendering, this might be the right time to investigate it if you're interested. My thoughts on what this could look like:

- We could provide almost infinite customization options. I think that rather than going down this path, we should enable a trap door that allows full control.
- The most obvious thing to try would be customization of the render() method. 
- Rather than extending each component and allowing render override, could we have an optional customRender(props) that the user could specify? Then inside of each component, we check for the that method and call it inside of render if it exists. 
- This way we can solve the issue for all custom scenarios. Every component can have an optional customRender()
- The tricky thing here is figuring out what the development workflow looks like I think. We'll have to see how to enable this from plain javascript as well as JSX and TSX projects. For plain js use would have to use [createElement](https://facebook.github.io/react/docs/react-api.html#createelement) from react. In typescript we need to make sure that the type signatures of those methods make sense. It would be best to automatically set the proptypes for customRender based on the signature of given component, but we need to figure out how that works. 

Thoughts?

**Comment by m-gagne** [comment]
Where this is coming from is building a POC where the search results is meta data and links to documents, something that aligns well to standard tabular layout. I also ran into the issue you mentioned around having headers (currently the way I have it working includes the header outside of the result, but then the count is placed below, then the results.

My first thought was to support two modes "tabular" or "block" based vs the infinitely customization. Though the ability to overwrite render could be interesting, or at the very least provide callbacks for the major aspects of it.

**Comment by EvanBoyle** [comment]
I can totally understand the desire for a tabular layout. I'm just not sure if specifying a flag for <div> vs <table> is the way to solve it. Ideally we can solve this problem in a general way that is reusable for all components that covers the gambit of customization scenarios. In general, since we're building a general purpose library I'm hesitant to add too many parameters (more parameters === additional complexity in the api for users) for niche scenarios that. As we add features we have to consider maintaining a reasonable api surface. 

Custom render is probably a big chunk of work. To unblock yourself for the POC, it would probably be much faster to just create a fork where you change the render method to use tables. I could look into custom render over the next two weeks, but if you're interested in taking a shot I'd be happy to help out with the design.

**Comment by m-gagne** [comment]
Agreed, too many optional parameters is confusing for sure. I wonder if something along the lines of `addCustomResults(htmlId, callback)` where the `callback` receives as the only parameters `this.props` from `render()` would be a viable solution. It's 100% customizable (though perhaps not the most user friendly, but would allow for any type of listing result (tables, sorted lists, unsorted lists, block elements, etc.)

**Comment by EvanBoyle** [comment]
I think this is a problem that generalizes to all components, not just results. Let's try to provide a uniform way to solve this problem that generalizes to all components. With that in mind I think that an optional paramteter to specify customRender(props) makes sense.
